spd/eval.py

```diff
@@ -755,6 +756,165 @@ def compute(self) -> Mapping[str, float]:
+            ref_flat = ref_rank_one.reshape(C, -1)
+
+            # Compute cosine similarities between all pairs
+            current_norm = F.normalize(current_flat, p=2, dim=1)
+            ref_norm = F.normalize(ref_flat, p=2, dim=1)
+
+            cosine_sim_matrix = einops.einsum(
+                current_norm, ref_norm, "C_alive d_in_d_out, C_ref d_in_d_out -> C_alive C_ref"
+            )
+            # Take the abs of the cosine similarity matrix
+            cosine_sim_matrix = cosine_sim_matrix.abs()
+
+            # Find max abs cosine similarity for each current component
+            max_similarities = cosine_sim_matrix.max(dim=1).values
+            similarities[f"mean_max_abs_cosine_sim/{layer_name}"] = max_similarities.mean().item()
+            similarities[f"max_abs_cosine_sim_std/{layer_name}"] = max_similarities.std().item()
+            similarities[f"max_abs_cosine_sim_min/{layer_name}"] = max_similarities.min().item()
+            similarities[f"max_abs_cosine_sim_max/{layer_name}"] = max_similarities.max().item()
+
+        # Compute a metrics across all model components for each type of metric
+        # First get the metric names by stripping away the layer name
+        metric_names = [
+            "mean_max_abs_cosine_sim",
+            "max_abs_cosine_sim_std",
+            "max_abs_cosine_sim_min",
+            "max_abs_cosine_sim_max",
+        ]
+
+        for metric_name in metric_names:
+            # Go through all layers and get the average of the metric
+            values = [
+                similarities[f"{metric_name}/{layer_name}"] for layer_name in self.model.components
+            ]
+            similarities[f"{metric_name}/all_layers"] = sum(values) / len(values)
+
+        return similarities
+
+    @override
+    def watch_batch(
+        self,
+        batch: Int[Tensor, "..."] | Float[Tensor, "..."],
+        target_out: Float[Tensor, "... vocab"],
+        ci: dict[str, Float[Tensor, "... C"]],
+    ) -> None:
+        n_tokens = next(iter(ci.values())).shape[:-1].numel()
+        self.n_tokens += n_tokens
+
+        for module_name, ci_vals in ci.items():
+            active_components = ci_vals > self.config.ci_alive_threshold
+            n_activations_per_component = reduce(active_components, "... C -> C", "sum")
+            self.component_activation_counts[module_name] += n_activations_per_component
+
+    @override
+    def compute(self) -> Mapping[str, float]:
+        """Compute the geometric similarity metrics"""
+
+        activation_densities = {
+            module_name: self.component_activation_counts[module_name] / self.n_tokens
+            for module_name in self.model.components
+        }
+
+        if self._computed_this_eval:
+            return {}
+
+        try:
+            similarities = self._compute_subcomponent_geometric_similarities(activation_densities)
+            self._computed_this_eval = True
+            return similarities
+        except Exception as e:
```

Comment:
I'd only have this try/except if it's possible for it to fail late in a run. If it will fail at step 0, I'd prefer to just have it raise an error immediately and let the user update it.

Relatedly, if keeping this, I think you should catch the specific exceptions you expect to see.

Reply:
(guess this is no longer relevant if making this a script rather than an eval. But what I mean was to catch e.g. ValueError instead of any Exception.